File: seulgi/2024-06/P258705_산모양타일링.py
def solution(n, tops):
    a = [0] * (n+1)
    b = [0] * (n+1)
    a[0] = 0
    b[0] = 1
    
    for i in range(1, n+1):
        if tops[i-1]:
            a[i] = (a[i-1] + b[i-1]) % 10007
            b[i] = (2 * a[i-1] + 3 * b[i-1]) % 10007
        else:
            a[i] = (a[i-1] + b[i-1]) % 10007
            b[i] = (a[i-1] + 2 * b[i-1]) % 10007
    return (a[n] + b[n]) % 10007

File: sseni/2024-06/P258705_산모양타일링.java
package com.example.algorithm.study;

public class P258705_산모양타일링 {
    public static int solution(int n, int[] tops) {
        int[] a = new int[n + 1];
        int[] b = new int[n + 1];

        a[1] = 1;
        if (tops[0] == 1) b[1] = 3;
        else b[1] = 2;

        for (int i = 2; i <= n; i++) {
            a[i] = (a[i - 1] + b[i - 1]) % 10007;
            if (tops[i - 1] == 1) {
                b[i] = (a[i - 1] * 2 + b[i - 1] * 3) % 10007;
            } else {
                b[i] = (a[i - 1] + b[i - 1] * 2) % 10007;
            }
        }
        return (a[n] + b[n]) % 10007;
    }

    public static void main(String[] args) {
        System.out.println(solution(4, new int[]{1, 1, 0, 1}));
    }
}
